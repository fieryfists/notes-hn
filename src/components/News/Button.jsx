const Button = (props) => {
    const { onClick, children, className } = props;
    return (
      <button type="button" onClick={onClick} className={className}>
        {children}
      </button>
    );
  };

  const Loading = () => {
    return (
      <div>
        <i className="fa fa-spinner" aria-hidden="true"></i>
      </div>
    );
  };

  const withLoading =
    (Component) =>
    ({ isLoading, ...props }) => {
      return isLoading ? <Loading /> : <Component {...props} />;
    };

  const ButtonWithLoading = withLoading(Button);

  export default Button;

  export { ButtonWithLoading };