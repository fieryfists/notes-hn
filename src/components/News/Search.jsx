const Search = (props) => {
  const { value, onChange, onSubmit, searchKey } = props;
  return (
    <form onSubmit={onSubmit}>
      <input
        type="text"
        value={value}
        onChange={onChange}
        placeholder="Search Here"
      />
      <button type="submit">
        <i className="fa fa-search" aria-hidden="true"></i>
      </button>
      &nbsp;
      <span className="rightSearch">
        {searchKey ? (
          <span>
            Showing stories about <span className="subject">{searchKey}</span>
          </span>
        ) : (
          <span>
            Showing <span className="subject">all</span> stories
          </span>
        )}
      </span>
    </form>
  );
};

export default Search;