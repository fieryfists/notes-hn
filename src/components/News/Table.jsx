import Sort from "./Sort";
import ModalWindow from "./ModalWindow";
import { sortBy } from "lodash";

const Table = (props) => {
  const { list, sortKey, onSort, isSortReverse, searchKey, onSave } = props;

  const SORTS = {
    NONE: (list) => list,
    TITLE: (list) => sortBy(list, "title"),
    AUTHOR: (list) => sortBy(list, "author"),
    COMMENTS: (list) => sortBy(list, "num_comments").reverse(),
    POINTS: (list) => sortBy(list, "points").reverse(),
  };

  const sortedList = SORTS[sortKey](list);

  const reverseSortedList = isSortReverse
    ? SORTS[sortKey](list).reverse()
    : sortedList;

  return (
    <div className="table">
      <div className="table-header">
        <span style={{ width: "30%" }}>
          <Sort sortKey={"TITLE"} onSort={onSort} activeSortKey={sortKey}>
            Title
          </Sort>
        </span>
        <span style={{ width: "30%" }}>
          <Sort sortKey={"AUTHOR"} onSort={onSort} activeSortKey={sortKey}>
            Author
          </Sort>
        </span>
        <span style={{ width: "15%" }}>
          <Sort sortKey={"COMMENTS"} onSort={onSort} activeSortKey={sortKey}>
            Comments
          </Sort>
        </span>
        <span style={{ width: "15%" }}>
          <Sort sortKey={"POINTS"} onSort={onSort} activeSortKey={sortKey}>
            Points
          </Sort>
        </span>
        <span style={{ width: "10%" }}>
          <Sort sortKey={"POINTS"} onSort={onSort} activeSortKey={sortKey}>
            Points
          </Sort>
        </span>
      </div>
      {reverseSortedList
        .map((item) => (
          <div key={item.objectID} className="table-row">
            <span style={{ width: "30%" }}>
              <a href={item.url} className="subject">
                {item.title}
              </a>
            </span>
            <span style={{ width: "30%" }}>{item.author}</span>
            <span style={{ width: "15%" }}>{item.num_comments}</span>
            <span style={{ width: "15%" }}>{item.points}</span>
            <span style={{ width: "10%" }}>
              <ModalWindow
                onSave={onSave}
                searchKey={searchKey}
                resultQuery={item}
              />
            </span>
          </div>
        ))}
    </div>
  );
};

export default Table;
