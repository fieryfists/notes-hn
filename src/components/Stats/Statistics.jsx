const Statistics = ({ queries, isVisible }) => {
  const getLastWeekHits = (query) => {
    const today = new Date().getTime();
    const startWeek = new Date(today - 7 * 24 * 60 * 60 * 1000);

    const rep = queries.filter(
      (item) =>
        Date.parse(item.creationDate) >= startWeek &&
        Date.parse(item.creationDate) <= today &&
        Date.parse(query.creationDate) >= startWeek &&
        Date.parse(query.creationDate) <= today &&
        item.query === query.query
    );

    const lastWeekHits = rep.reduce((sum, current, _, { length }) => {
      return sum + current.hits;
    }, 0);

    return Math.round(lastWeekHits / rep.length) || 0;
  };

  return (
    <div className={`table ${isVisible ? "visible" : "hidden"}`}>
      <div className="table-header">
        <span style={{ width: "50%" }}>SEARCH KEY</span>
        <span style={{ width: "15%" }}>CREATION DATE</span>
        <span style={{ width: "15%" }}>HITS/DAY</span>
        <span style={{ width: "20%" }}>HITS/LAST WEEK</span>
      </div>

      {queries.map((item) => (
        <div key={item.id} className="table-row">
          <span style={{ width: "50%" }}>
            <a href={item.url} className="subject">
              {item.query}
            </a>
          </span>
          <span style={{ width: "15%" }}>{item.creationDate.slice(0, 10)}</span>
          <span style={{ width: "15%" }}>{item.hits}</span>
          <span style={{ width: "20%" }}>{getLastWeekHits(item)}</span>
        </div>
      ))}
    </div>
  );
};

export default Statistics;
