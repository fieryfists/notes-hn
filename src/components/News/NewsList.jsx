import React, { Component } from "react";
import "font-awesome/css/font-awesome.min.css";
import Search from "./Search";
import Table from "./Table";
import { ButtonWithLoading } from "./Button";

const DEFAULT_QUERY = "";
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = 20;
const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

class NewsList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      results: null,
      query: DEFAULT_QUERY,
      page: DEFAULT_PAGE,
      searchKey: "",
      isLoading: false,
      sortKey: "NONE",
      isSortReverse: false,
    };

    this.onSave = this.props.onSave;
    this.afterSearch = this.props.afterSearch;

    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.onSort = this.onSort.bind(this);
  }

  setSearchTopStories(result) {
    const { hits, page, nbHits } = result;
    const { searchKey } = this.state;
    const oldHits = page === 0 ? [] : this.state.results[searchKey].hits;
    const updatedHits = [...oldHits, ...hits];

    if (searchKey.length > 0) this.afterSearch(searchKey, nbHits);

    this.setState({
      results: {
        ...this.state.results,
        [searchKey]: { hits: updatedHits, page },
      },
      isLoading: false,
    });
  }

  fetchSearchTopStories(query, page) {
    this.setState({ isLoading: true });
    fetch(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${query}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then((response) => response.json())
      .then((result) => this.setSearchTopStories(result));
  }

  onSearchChange(event) {
    this.setState({ query: event.target.value, page: 0 });
  }

  onSearchSubmit(event) {
    const { query, page } = this.state;
    this.setState({ searchKey: query });
    if (this.needsToSearchTopStories(query)) {
      this.fetchSearchTopStories(query, page);
    }
    event.preventDefault();
  }

  needsToSearchTopStories(query) {
    return !this.state.results[query];
  }

  onSort(sortKey) {
    const isSortReverse = sortKey === this.state.sortKey ? true : false;
    this.setState({ sortKey: sortKey, isSortReverse: isSortReverse });
  }

  componentDidMount() {
    const { query, page } = this.state;
    this.setState({ searchKey: query });
    this.fetchSearchTopStories(query, page);
  }

  componentDidUpdate(prevProps) {
    const { defaultQuery } = this.props;
    if (prevProps.defaultQuery !== defaultQuery) {
      if (defaultQuery) {
        this.setState({ searchKey: defaultQuery, query: defaultQuery, page: 0, results: {} });
        this.fetchSearchTopStories(defaultQuery, DEFAULT_PAGE);
      }
    }
  }

  render() {
    const { results, query, searchKey, isLoading, sortKey, isSortReverse } =
      this.state;
    const page = results && results[searchKey] ? results[searchKey].page : 0;
    const list = results && results[searchKey] ? results[searchKey].hits : [];

    const { isVisible } = this.props;
    return (
      <div className={`page ${isVisible ? "visible" : "hidden"}`}>
        <div className="interactions">
          <Search
            value={query}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            afterSearch={this.afterSearch}
            searchKey={searchKey}
          />
        </div>
        <Table
          list={list}
          sortKey={sortKey}
          onSort={this.onSort}
          isSortReverse={isSortReverse}
          searchKey={searchKey}
          onSave={this.onSave}
        />
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
          >
            More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }
}

export default NewsList;
