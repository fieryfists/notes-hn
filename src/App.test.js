import React from "react";
import ReactDOM from "react-dom";
import rendered from "react-test-renderer";
import NewsList from "./components/News/NewsList";
import Statistics from "./components/Stats/Statistics";
import Search from './components/News/Search';
import Button from './components/News/Button';
import Table from './components/News/Table';

describe("Search", () => {
  it("renders", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Search>Search</Search>, div);
  });

  test("snapshots", () => {
    const component = rendered.create(<Search>Search</Search>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Button", () => {
  it("renders", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Button>More</Button>, div);
  });

  test("snapshots", () => {
    const component = rendered.create(<Button>More</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Table", () => {
  const props = {
    list: [
      { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
      { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" },
    ],
    sortKey: "NONE",
  };

  it("renders", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Table {...props} />, div);
  });

  test("snapshots", () => {
    const component = rendered.create(<Table {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("NewsList", () => {
  it("renders", () => {
    const div = document.createElement("div");
    ReactDOM.render(<NewsList />, div);
  });
});

it("api call returns a string", async () => {
  const getData = () => {
    return fetch(
      "https://hn.algolia.com/api/v1/search?query=&page=0&hitsPerPage=1"
    ).then((res) => res.json());
  };

  const expected = {
    hits: expect.arrayContaining([
      expect.objectContaining({
        created_at: expect.any(String),
        title: expect.any(String),
        url: expect.any(String),
        author: expect.any(String),
        points: expect.any(Number),
        num_comments: expect.any(Number),
        relevancy_score: expect.any(Number),
        _tags: expect.any(Array),
        objectID: expect.any(String),
        _highlightResult: expect.any(Object),
      }),
    ]),
    nbHits: expect.any(Number),
    page: expect.any(Number),
    nbPages: expect.any(Number),
    query: expect.any(String),
  };
  let actual = await getData();

  expect(actual.hits).toHaveLength(1);
  expect(actual).toMatchObject(expected);
});

describe("Statistics", () => {
  const props = {
    queries: [
      {
        creationDate: "2021-10-23T12:48:12.353Z",
        hits: 11413,
        id: 0,
        query: "fdfs",
      },
      {
        creationDate: "2021-10-22T12:48:12.353Z",
        hits: 11412,
        id: 1,
        query: "fdfs",
      },
      {
        creationDate: "2021-10-21T12:48:12.353Z",
        hits: 11411,
        id: 2,
        query: "fdfs",
      },
      {
        creationDate: "2021-09-20T12:48:12.353Z",
        hits: 11410,
        id: 3,
        query: "fdfs",
      },
    ],
    isVisible: true,
  };

  it("renders", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Statistics {...props} />, div);
  });

  test("snapshots", () => {
    const component = rendered.create(<Statistics {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
