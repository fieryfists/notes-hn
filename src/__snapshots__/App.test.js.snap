// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots 1`] = `
<button
  type="button"
>
  More
</button>
`;

exports[`Search snapshots 1`] = `
<form>
  <input
    placeholder="Search Here"
    type="text"
  />
  <button
    type="submit"
  >
    <i
      aria-hidden="true"
      className="fa fa-search"
    />
  </button>
  Â 
  <span
    className="rightSearch"
  >
    <span>
      Showing 
      <span
        className="subject"
      >
        all
      </span>
       stories
    </span>
  </span>
</form>
`;

exports[`Statistics snapshots 1`] = `
<div
  className="table visible"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      SEARCH KEY
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      CREATION DATE
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      HITS/DAY
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      HITS/LAST WEEK
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        className="subject"
      >
        fdfs
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2021-10-23
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      11413
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      11412
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        className="subject"
      >
        fdfs
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2021-10-22
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      11412
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      11412
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        className="subject"
      >
        fdfs
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2021-10-21
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      11411
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      11412
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        className="subject"
      >
        fdfs
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2021-09-20
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      11410
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      0
    </span>
  </div>
</div>
`;

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <a
        className="subject"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <span>
        <button
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="fa fa-heart"
          />
        </button>
      </span>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <a
        className="subject"
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <span>
        <button
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="fa fa-heart"
          />
        </button>
      </span>
    </span>
  </div>
</div>
`;
